File: crates/ubaa-core/src/domain/ygdk.rs
use std::fmt;

use serde::{Deserialize, Serialize};

use super::ActionEligibility;

/// 阳光打卡提交的稳定目标。
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkSubmitTarget {
    pub classify_id: i32,
    pub item_id: i32,
}

/// 阳光打卡项目。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkItem {
    pub item_id: i32,
    pub name: String,
    pub kind: Option<i32>,
    pub sort: Option<i32>,
    pub submit_eligibility: ActionEligibility,
    pub submit_target: Option<YgdkSubmitTarget>,
}
/// 阳光打卡学期统计。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkTermSummary {
    pub term_id: Option<i32>,
    pub term_name: Option<String>,
    pub term_count: i32,
    pub term_target: Option<i32>,
    pub week_count: Option<i32>,
    pub week_target: Option<i32>,
    pub month_count: Option<i32>,
    pub month_target: Option<i32>,
    pub day_count: Option<i32>,
    pub good_count: Option<i32>,
}
/// 阳光打卡概览。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkOverview {
    pub summary: YgdkTermSummary,
    pub classify_id: i32,
    pub classify_name: String,
    pub default_item_id: i32,
    pub default_item_name: String,
    pub items: Vec<YgdkItem>,
}
/// 阳光打卡记录。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkRecord {
    pub record_id: i32,
    pub item_id: Option<i32>,
    pub item_name: Option<String>,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    pub place: Option<String>,
    pub images: Vec<String>,
    pub is_open: bool,
    pub state: Option<i32>,
    pub created_at: Option<String>,
    pub created_at_label: Option<String>,
}
/// 阳光打卡记录分页。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkRecordsPage {
    pub content: Vec<YgdkRecord>,
    pub total: i32,
    pub page: i32,
    pub size: i32,
    pub has_more: bool,
}
/// 阳光打卡图片上传。图片字节只在一次请求内存中存在，不写入会话或输出。
#[derive(Clone, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkPhotoUpload {
    #[serde(skip_serializing)]
    pub bytes: Vec<u8>,
    #[serde(skip_serializing)]
    pub file_name: String,
    #[serde(skip_serializing)]
    pub mime_type: String,
}
impl YgdkPhotoUpload {
    pub(crate) fn normalized_mime_type(&self) -> Option<&str> {
        let mime_type = self.mime_type.as_str();
        let subtype = mime_type.strip_prefix("image/")?;
        (mime_type == mime_type.trim()
            && !subtype.is_empty()
            && subtype.is_ascii()
            && subtype.bytes().all(is_http_token_character))
        .then_some(mime_type)
    }
}

fn is_http_token_character(value: u8) -> bool {
    value.is_ascii_alphanumeric()
        || matches!(
            value,
            b'!' | b'#'
                | b'$'
                | b'%'
                | b'&'
                | b'\''
                | b'*'
                | b'+'
                | b'-'
                | b'.'
                | b'^'
                | b'_'
                | b'`'
                | b'|'
                | b'~'
        )
}
impl fmt::Debug for YgdkPhotoUpload {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("YgdkPhotoUpload")
            .field("bytes", &format_args!("[{} bytes]", self.bytes.len()))
            .field("file_name", &"[REDACTED]")
            .field(
                "mime_type",
                &self.normalized_mime_type().unwrap_or("[INVALID]"),
            )
            .finish()
    }
}
/// 阳光打卡提交请求。
#[derive(Clone, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkClockinSubmitRequest {
    pub target: YgdkSubmitTarget,
    pub start_time: String,
    pub end_time: String,
    pub place: Option<String>,
    pub share_to_square: bool,
    pub photo: YgdkPhotoUpload,
}
impl fmt::Debug for YgdkClockinSubmitRequest {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("YgdkClockinSubmitRequest")
            .field("target", &self.target)
            .field("start_time", &"[已隐藏]")
            .field("end_time", &"[已隐藏]")
            .field("place", &self.place.as_ref().map(|_| "[已隐藏]"))
            .field("share_to_square", &self.share_to_square)
            .field("photo", &self.photo)
            .finish()
    }
}
/// 阳光打卡提交前由 Core fresh 复核形成的规范化结果。
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct YgdkSubmitPreflight {
    pub request: YgdkClockinSubmitRequest,
    pub item_name: String,
}
/// 阳光打卡提交结果。
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YgdkClockinSubmitResult {
    pub success: bool,
    pub message: String,
    pub record_id: Option<i32>,
}

File: packages/ubaa_app/lib/src/bridge/read/ygdk.dart
part of '../bridge_backend.dart';

Future<FeatureResult> _loadYgdkOverviewOnRoute(
  BridgeBackend backend, {
  required ConnectionMode route,
}) async {
  try {
    final result = await backend.client.ygdkOverviewOnRoute(
      route: _toBridgeConnectionMode(route),
    );
    return _mapYgdkOverviewResult(
      result.data,
      _toConnectionMode(result.pinnedRoute),
    );
  } on BridgeError catch (error) {
    throw _mapError(error);
  }
}

Future<FeatureResult> _loadYgdkRecordsOnRoute(
  BridgeBackend backend, {
  required ConnectionMode route,
  required int page,
  required int size,
}) async {
  if (page <= 0 || size <= 0) {
    throw const BackendException(UbaaErrorCode.invalidInput);
  }
  try {
    final result = await backend.client.ygdkRecordsOnRoute(
      route: _toBridgeConnectionMode(route),
      page: page,
      size: size,
    );
    return _mapYgdkRecordsResult(
      result.data,
      _toConnectionMode(result.pinnedRoute),
    );
  } on BridgeError catch (error) {
    throw _mapError(error);
  }
}

Future<FeatureResult> _loadYgdkFeature(
  BridgeBackend backend,
  FeatureId feature,
  FeatureQuery query, {
  bool includeRecords = false,
}) async {
  final client = backend.client;
  switch (feature) {
    case FeatureId.ygdk:
      switch (query.view) {
        case FeatureQueryView.summary:
          final result = await client.ygdkOverview();
          YgdkHomeRecords? records;
          if (includeRecords) {
            final page = query.page <= 0 ? 1 : query.page;
            final size = query.size.clamp(1, 100);
            try {
              final response = await client.ygdkRecordsOnRoute(
                route: result.route.resolvedRoute,
                page: page,
                size: size,
              );
              if (response.pinnedRoute != result.route.resolvedRoute) {
                throw const BackendException(UbaaErrorCode.internalError);
              }
              records = YgdkHomeRecords(
                page: response.data.page,
                size: response.data.size,
                total: response.data.total,
                hasMore: response.data.hasMore,
                content: List.unmodifiable(
                  response.data.content.map(_ygdkRecordPresentation),
                ),
              );
            } on BridgeError catch (error) {
              // 认证失效交给原控制器；其他局部失败保留概要，不伪造空记录。
              if (error.kind == BridgeErrorKind.authentication) rethrow;
              records = YgdkHomeRecords(
                page: page,
                size: size,
                errorCode: _typedErrorCode(error.code),
              );
            }
          }
          return _mapYgdkOverviewResult(
            result.data,
            _toConnectionMode(result.route.resolvedRoute),
            records: records,
          );
        case FeatureQueryView.ygdkRecords:
          final page = query.page <= 0 ? 1 : query.page;
          final size = query.size.clamp(1, 100);
          final result = await client.ygdkRecords(page: page, size: size);
          return _mapYgdkRecordsResult(
            result.data,
            _toConnectionMode(result.route.resolvedRoute),
          );
        case FeatureQueryView.libbookAreas:
        case FeatureQueryView.bykcDetail:
        case FeatureQueryView.bykcProfile:
        case FeatureQueryView.bykcChosenCourses:
        case FeatureQueryView.bykcStatistics:
        case FeatureQueryView.scheduleToday:
        case FeatureQueryView.scheduleTerms:
        case FeatureQueryView.scheduleWeeks:
        case FeatureQueryView.scheduleWeek:
        case FeatureQueryView.examArranged:
        case FeatureQueryView.examNotArranged:
        case FeatureQueryView.gradesScored:
        case FeatureQueryView.gradesMissing:
        case FeatureQueryView.evaluationPending:
        case FeatureQueryView.libbookAreaDetail:
        case FeatureQueryView.libbookSeats:
        case FeatureQueryView.libbookBookings:
        case FeatureQueryView.cgyyPurposeTypes:
        case FeatureQueryView.cgyyDayInfo:
        case FeatureQueryView.cgyyOrders:
        case FeatureQueryView.cgyyOrderDetail:
        case FeatureQueryView.cgyyLockCode:
        case FeatureQueryView.spocDetail:
        case FeatureQueryView.judgeDetail:
        case FeatureQueryView.judgeBatchDetails:
        case FeatureQueryView.signinPending:
        case FeatureQueryView.signinCompleted:
          throw const BackendException(UbaaErrorCode.invalidInput);
      }
    default:
      throw StateError('unexpected feature: $feature');
  }
}

FeatureResult _mapYgdkOverviewResult(
  BridgeYgdkOverview data,
  ConnectionMode resolvedRoute, {
  YgdkHomeRecords? records,
}) {
  final itemIdCounts = <int, int>{};
  for (final item in data.items) {
    itemIdCounts.update(item.itemId, (count) => count + 1, ifAbsent: () => 1);
  }
  final details = data.items
      .map(
        (item) => FeatureDetail(
          title: item.name,
          presentation: YgdkItemPresentation(
            itemId: item.itemId,
            name: item.name,
            kind: item.kind,
            sort: item.sort,
          ),
          fields: _compactFields(<FeatureField?>[
            _field('项目编号', '${item.itemId}'),
            item.kind == null ? null : _field('类型', '${item.kind}'),
          ]),
          actions: _ygdkSubmitActions(data, item, itemIdCounts),
        ),
      )
      .toList(growable: false);
  final summary = data.summary.termTarget == null
      ? '已打卡 ${data.summary.termCount} 次'
      : '学期进度 ${data.summary.termCount}/${data.summary.termTarget}';
  return FeatureResult.success(
    summary: summary,
    overview: YgdkOverview(
      termId: data.summary.termId,
      termName: data.summary.termName,
      termCount: data.summary.termCount,
      termTarget: data.summary.termTarget,
      weekCount: data.summary.weekCount,
      weekTarget: data.summary.weekTarget,
      monthCount: data.summary.monthCount,
      monthTarget: data.summary.monthTarget,
      dayCount: data.summary.dayCount,
      goodCount: data.summary.goodCount,
      classifyId: data.classifyId,
      classifyName: data.classifyName,
      defaultItemId: data.defaultItemId,
      defaultItemName: data.defaultItemName,
      records: records,
    ),
    details: details,
    resolvedRoute: resolvedRoute,
  );
}

FeatureResult _mapYgdkRecordsResult(
  BridgeYgdkRecordsPage data,
  ConnectionMode resolvedRoute,
) {
  final details = data.content
      .map(
        (item) => FeatureDetail(
          title: item.itemName ?? '打卡记录 ${item.recordId}',
          subtitle: item.createdAtLabel ?? item.createdAt,
          presentation: _ygdkRecordPresentation(item),
          fields: _compactFields(<FeatureField?>[
            _field('记录编号', '${item.recordId}'),
            _field('开始时间', item.startTime),
            _field('结束时间', item.endTime),
            _field('地点', item.place),
            _field('公开状态', item.isOpen ? '公开' : '不公开'),
            _field('图片数量', '${item.imageCount}'),
          ]),
        ),
      )
      .toList(growable: false);
  return _countResult(
    data.content.length,
    '条打卡记录',
    details: details,
    pagination: _pagination(
      page: data.page,
      size: data.size,
      total: data.total,
      hasMore: data.hasMore,
    ),
    resolvedRoute: resolvedRoute,
  );
}

List<FeatureAction> _ygdkSubmitActions(
  BridgeYgdkOverview overview,
  BridgeYgdkItem item,
  Map<int, int> itemIdCounts,
) {
  if (overview.classifyId <= 0 ||
      overview.classifyName.trim().isEmpty ||
      item.itemId <= 0 ||
      item.name.trim().isEmpty ||
      itemIdCounts[item.itemId] != 1) {
    return const <FeatureAction>[];
  }
  final target = item.submitTarget;
  if (item.submitEligibility != BridgeActionEligibility.allowed ||
      target == null ||
      target.classifyId != overview.classifyId ||
      target.itemId != item.itemId ||
      target.classifyId <= 0 ||
      target.itemId <= 0) {
    return const <FeatureAction>[];
  }
  return <FeatureAction>[
    YgdkSubmitAction(
      classifyId: target.classifyId,
      itemId: target.itemId,
      eligibility: _toYgdkActionEligibility(item.submitEligibility),
    ),
  ];
}

ActionEligibility _toYgdkActionEligibility(
  BridgeActionEligibility eligibility,
) => switch (eligibility) {
  BridgeActionEligibility.allowed => ActionEligibility.allowed,
  BridgeActionEligibility.denied => ActionEligibility.denied,
  BridgeActionEligibility.unknown => ActionEligibility.unknown,
};

YgdkRecordPresentation _ygdkRecordPresentation(BridgeYgdkRecord item) =>
    YgdkRecordPresentation(
      recordId: item.recordId,
      itemId: item.itemId,
      itemName: item.itemName,
      startTime: item.startTime,
      endTime: item.endTime,
      place: item.place,
      imageCount: item.imageCount,
      isOpen: item.isOpen,
      state: item.state,
      createdAt: item.createdAt,
      createdAtLabel: item.createdAtLabel,
    );
